pensieve/config.py

```diff
@@ -0,0 +1,160 @@
+import enum
+from typing import Dict, List, Optional
+
+import attr
+import cattr
+import mozanalysis.metrics
+import mozanalysis.metrics.desktop
+import mozanalysis.segments
+
+import pensieve.experimenter
+
+"""
+Spec objects are direct representations of the configuration and contain unresolved references
+to metrics and data sources.
+
+Calling .resolve(config_spec) on a Spec object produces a concrete resolved Conf class.
+"""
+
+
+_converter = cattr.Converter()
+
+
+@enum.unique
+class Platform(enum.Enum):
+    DESKTOP = "desktop"
+    FENIX = "fenix"
+
+
+@attr.s(auto_attribs=True)
+class MetricReference:
+    name: str
+
+    def resolve(self, spec: "ConfigurationSpec") -> mozanalysis.metrics.Metric:
+        search = None
+        if spec.experiment.platform == Platform.DESKTOP:
+            search = mozanalysis.metrics.desktop
+
+        attr = None
+        if search and hasattr(search, self.name):
+            attr = getattr(search, self.name)
+        if self.name in spec.metrics.definitions:
+            attr = spec.metrics.definitions[self.name].resolve(spec)
+        if isinstance(attr, mozanalysis.metrics.Metric):
+            return attr
+
+        raise ValueError(f"Could not find the metric {self.name}")
+
+
+@attr.s(auto_attribs=True)
+class DataSourceReference:
+    name: str
+
+    def resolve(self, cfg: "ConfigurationSpec") -> mozanalysis.metrics.DataSource:
+        raise NotImplementedError
+
+
+@attr.s(auto_attribs=True)
+class ExperimentSpec:
+    def sync(self, experiment: pensieve.experimenter.Experiment):
+        raise NotImplementedError
+
+
+@attr.s(auto_attribs=True)
+class MetricDefinition:
+    name: str  # implicit in configuration
+    select_expression: str
+    data_source: DataSourceReference
+
+    def resolve(self, spec: "ConfigurationSpec") -> mozanalysis.metrics.Metric:
+        return mozanalysis.metrics.Metric(
+            name=self.name,
+            data_source=self.data_source.resolve(spec),
+            select_expr=self.select_expression,
+        )
+
+
+@attr.s(auto_attribs=True)
+class MetricsConfiguration:
+    daily: List[mozanalysis.metrics.Metric]
+    weekly: List[mozanalysis.metrics.Metric]
+    overall: List[mozanalysis.metrics.Metric]
+
+
+@attr.s(auto_attribs=True)
+class MetricsSpec:
+    daily: List[MetricReference]
+    weekly: List[MetricReference]
+    overall: List[MetricReference]
+
+    definitions: Dict[str, MetricDefinition]
+
+    @classmethod
+    def from_dict(cls, d: dict) -> "MetricsSpec":
+        params = {}
+        for k in ("daily", "weekly", "overall"):
+            params[k] = d.get(k, [])
+            del d[k]
```

Comment:
Is this side-effect necessary?

Reply:
oh, oops, mutating the input is quite rude of me